receiver/prometheusexecreceiver/subprocessmanager/manager.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package subprocessmanager
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/kballard/go-shellquote"
+	"go.uber.org/zap"
+)
+
+// Process struct holds all the info needed to instantiate a subprocess
+type Process struct {
+	Command string
+	Port    int
+	Env     []EnvConfig
+}
+
+const (
+	// HealthyProcessTime is the default time a process needs to stay alive to be considered healthy
+	HealthyProcessTime time.Duration = 30 * time.Minute
+	// HealthyCrashCount is the amount of times a process can crash (within the healthyProcessTime) before being considered unstable - it may be trying to find a port
+	HealthyCrashCount int = 3
+	// delayMutiplier is the factor by which the delay scales
+	delayMultiplier float64 = 2.0
+	// initialDelay is the initial delay before a process is restarted
+	initialDelay time.Duration = 1 * time.Second
+)
+
+// Run will start the process and keep track of running time
+func (proc *Process) Run(logger *zap.Logger) (time.Duration, error) {
```

Comment:
There is no way to gracefully kill the subprocess from what I can tell when the collector shuts down.  For stateless exporter process graceful shutdown probably doesn't matter so much but you still want to make sure that the process actually does get shutdown if the collector dies.  I believe on Linux at least this is not guaranteed unless you do something like [this](https://github.com/open-telemetry/opentelemetry-collector/pull/1381/files#diff-3b0cfd94cd2dd343186bdea637fadcd2R25).  You can test it easily by running the collector with this receiver, doing a kill on the collector, and seeing if the exporter process also exits, or if it hangs around.

Reply:
I see what you mean. I was gracefully stopping it since the beginning but killing it does make the process hang around. I'll modify this, thanks!